RA-1902: Add An Icon with Functionality of Showing Password on Login Page

jnsereko opened the pull request:
Added An Icon with Functionality of Show Password on Login Page

Issue link:   https://issues.openmrs.org/browse/RA-1902

jnsereko commented:
cc @sherrif10 
I need to make the icon be in one line with the password

jnsereko commented on omod/src/main/webapp/pages/login.gsp:
```diff
@@ -148,7 +157,10 @@
                                     ${ ui.message("referenceapplication.login.password") }:
                                 </label>
-                                <input class="form-control form-control-sm form-control-lg form-control-md" id="password" type="password" name="password" placeholder="${ ui.message("referenceapplication.login.password.placeholder") }"/>
+                                <span>
```
`

jnsereko commented on omod/src/main/webapp/pages/login.gsp:
```diff
@@ -148,7 +157,10 @@
                                 </label>
-                                <input class="form-control form-control-sm form-control-lg form-control-md" id="password" type="password" name="password" placeholder="${ ui.message("referenceapplication.login.password.placeholder") }"/>
+                                <span>
+	                                <input class="form-control form-control-sm form-control-lg form-control-md" id="password" type="password" name="password" placeholder="${ ui.message("referenceapplication.login.password.placeholder") }"/>
```
`<input style="display: inline" .........  />`

![image](https://user-images.githubusercontent.com/58003327/112871597-11b70c80-90c8-11eb-9a64-5b149ecb3e0f.png)

HerbertYiga replied:
@jnsereko what current screen shot dou you have  for these changes i see some here and at the ticket Id

jnsereko replied:
the one here are just used to show what happens when i change component css in chrome inspection.

sherrif10 replied:
Looks good @jnsereko  , can you try putting it inside the password tab, cc @ibacher th we had suggested of  catering  microsoft edge use case  could you please advise on this thanks .

jnsereko replied:
ooohhh working on it

jnsereko commented:
![Screenshot from 2021-07-19 17-04-07](https://user-images.githubusercontent.com/58003327/126172746-b9e00ac7-00de-46f6-852f-084be7b8c079.png)

I have put the icon in the password field and I think it is better this way

sherrif10 commented on omod/src/main/webapp/pages/login.gsp:
```diff
@@ -21,6 +21,15 @@
+        	jQuery(this).toggleClass("fa-eye fa-eye-slash");
+            var type = jQuery(this).hasClass("fa-eye-slash") ? "text" : "password";
+            jQuery("#password").attr("type", type);
+        });
```
The icon would look more clear when put inside the bars, However i remember @ibacher  commenting about this that we should cater for authorization

jnsereko replied:
>  inside the bars,

Hey @sherrif10 
Am sorry to ask but which bars are that 



> commenting about this that we should cater for authorization

Did @ibacher comment here, I am not seeing the comment.  I request you to send me the link here.

jnsereko commented on omod/src/main/webapp/pages/login.gsp:
```diff
@@ -21,6 +21,15 @@
+        	jQuery(this).toggleClass("fa-eye fa-eye-slash");
+            var type = jQuery(this).hasClass("fa-eye-slash") ? "text" : "password";
+            jQuery("#password").attr("type", type);
+        });
```
hey @ibacher 
I think the symbol **$** wasn't defined 
![image](https://user-images.githubusercontent.com/58003327/126450803-d223cfb2-4977-4cdf-8052-a1b35d4c5bfe.png)

so I have tried to change the variable name from 

> var $this = jQuery(this);

`var pwdIcon = jQuery(this);`

ibacher replied:
Oh... right. This is in a JSP so that `$this` gets read by Groovy and tries to interpolate the value, then it doesn't work because it would need to be, e.g., `${this}` and then it's whatever the `toString()` method of the Groovy page returns, so probably not what we need. 

Good job renaming the variable. I was just trying to replicate the thing that's sometimes done of prefixing a variable with `$` (usually how you can reference jQuery) to indicate it was already a jQuery object.

jnsereko replied:
thanks for the great explanation @ibacher 
I think this is now almost good.  cc @isears

jnsereko commented:
I have added some changes and I request for review
cc @isears @ibacher @sherrif10

isears commented on omod/src/main/webapp/pages/login.gsp:
```diff
@@ -155,8 +165,9 @@
                             <p class="left">
                                 <label for="password">
                                     ${ ui.message("referenceapplication.login.password") }:
+                                    <i class="far fa-eye" style="float: right" id="togglePassword"></i>
```
I think in-line styling (i.e. `style="Float: right"`) can get confusing, especially since we already have a stylesheet for the login page: https://github.com/openmrs/openmrs-module-referenceapplication/blob/db4f80bef85555ff9e540d971ea818cec106207d/omod/src/main/compass/sass/login.scss. Let's just refactor the styling directive into the stylesheet.

I also like @sherrif10 's suggestion of including the eye icon within the textbox. Seems like this could be achieved fairly easily with a negative margin like here: https://www.javascripttutorial.net/javascript-dom/javascript-toggle-password-visibility/

jnsereko replied:
> like @sherrif10 's suggestion of including the eye icon within the textbox

Yeah, @isears! I definitely agree with you!  However, I tried implementing this but the results were ugly.  
```
<p class="left">
   <label for="password">
         ${ ui.message("referenceapplication.login.password") }:
   </label>
   <input class="form-control form-control-sm form-control-lg form-control-md" id="password" type="password" name="password" placeholder="${ ui.message("referenceapplication.login.password.placeholder") }"/>
   <i class="far fa-eye" style="float: right" id="togglePassword"></i>
 </p>
```
Am going to retry again

jnsereko commented on omod/src/main/compass/sass/login.scss:
```diff
@@ -6,6 +6,13 @@
+      position: relative;
+      margin: -50px 10px;
+      float: right;
     }
```
@isears @ibacher when this is interpreted, it turns to 
```
#login-form p #togglePassword {
  position: relative;
  margin: -50px 10px;
  float: right;
}
```
which looks to work very fin, however, the server doesn't show the icon inside the password text field.  However, manual adding of the css from firefox inspect makes the goal attained

jnsereko commented:
hey @ibacher @isears 
I have just found out that if you need css to be loaded, you must run the SDK inside the home directory of the repository.  
it works fine 
![image](https://user-images.githubusercontent.com/58003327/126778461-384b1389-e2b8-40b0-a38c-7f223827d2e6.png)

sherrif10 commented:
Thanks @jnsereko  did you see build failures

ibacher commented:
@sherrif10  Unfortunately, those build failures are caused by the issue [discussed here](https://talk.openmrs.org/t/github-actions-builds-failing-on-ref-app-and-hfe-modules-and-probably-more-projects/33639/3) and aren't specific to this PR.

sherrif10 commented:
thanks @ibacher  My bad i had not seen the discussion, probably will look into the discussions

ibacher commented:
@jnsereko I updated `master` so that PRs should be able to be built by GitHub Actions again, but in the course of this, I moved the Sass files to `src/main/resources/styles`. Would you be able to update this PR to handle that? Thanks!

jnsereko commented:
hello @ibacher @isears @sherrif10 
I have made some changes.  I request for review

isears reviewed (commented):
@jnsereko 

When I test this locally, it seems like the rules from `login.scss` aren't being applied for some reason (Google Chrome):
![image](https://user-images.githubusercontent.com/7016326/126992775-207c43af-356a-4501-87f1-d45b8ce1658f.png)

Maybe you could try directly including the rules form `login.scss` using `ui.includeCss()` within `login.gsp` instead of the `@import "login";` statement in `referenceapplication.scss`?

jnsereko commented:
Hello @isears 
Both attempts are working on my side.  Am thinking of the possible reasons why this is not working on your side.

1. I removed the line `@import "login";` from the `referenceapplication.scss` file and CSS **didn't** load.  The icon `wasn't` inserted in the password field.

2. I added the line `@import "login"; to the `referenceapplication.scss` file and CSS **was** loaded. and the icon was displayed in the password field.

isears commented:
Ok, maybe I have some bad CSS cached somewhere, I will try again with a fresh server and let you know

jnsereko commented:
hey @isears!  Thanks for the review.
However, I have added the `login.css` file in the `login.gsp` file but everything also works fine.
```
<% ui.includeCss("login.css") %>
```
Maybe you can pull changes and try to see if it works on your side.
Thank you so much

isears reviewed (approved):
Ok this looks good to me now. I'll leave it open until Tuesday morning (UTC-4) for any other last-minute comments, then merge it in unless there are objections.

Thanks @jnsereko !

sherrif10 reviewed (approved):
Thanks @jnsereko
